server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java

```diff
@@ -0,0 +1,603 @@
+  }
+
+  @Override
+  public void unmarkSegmentToDrop(DataSegment dataSegment)
+  {
+    segmentsMarkedToDrop.remove(dataSegment);
+  }
+
+  @Override
+  public int getNumberOfSegmentsInQueue()
+  {
+    return segmentsToLoad.size();
+  }
+
+  @Override
+  public Set<DataSegment> getSegmentsMarkedToDrop()
+  {
+    return segmentsMarkedToDrop;
+  }
+
+  private class SegmentHolder
+  {
+    private final DataSegment segment;
+    private final DataSegmentChangeRequest changeRequest;
+    private final int type;
+    private final List<LoadPeonCallback> callbacks = Lists.newArrayList();
+
+    private SegmentHolder(
+        DataSegment segment,
+        int type,
+        LoadPeonCallback callback
+    )
+    {
+      this.segment = segment;
+      this.type = type;
+      this.changeRequest = (type == LOAD)
+                           ? new SegmentChangeRequestLoad(segment)
+                           : new SegmentChangeRequestDrop(segment);
+
+      if (callback != null) {
+        this.callbacks.add(callback);
+      }
+
+      if (type == LOAD) {
+        queuedSize.addAndGet(segment.getSize());
+      }
+    }
+
+    public DataSegment getSegment()
+    {
+      return segment;
+    }
+
+    public int getType()
+    {
+      return type;
+    }
+
+    public void addCallback(LoadPeonCallback newCallback)
+    {
+      synchronized (callbacks) {
+        if (newCallback != null) {
+          callbacks.add(newCallback);
+        }
+      }
+    }
+
+    public DataSegmentChangeRequest getChangeRequest()
+    {
+      return changeRequest;
+    }
+
+    AtomicReference<Future> cancellationFutureRef = new AtomicReference<>();
+
+    public void scheduleCancellationOnTimeout()
+    {
+      cancellationFutureRef.updateAndGet(
+          (curr) -> {
+            if (curr == null) {
+              return processingExecutor.schedule(
+                  () -> {
+                    synchronized (lock) {
+                      if (type == LOAD) {
+                        segmentsToLoad.remove(segment);
+                      } else {
+                        segmentsToDrop.remove(segment);
+                      }
+                      requestCompleted(
+                          true,
+                          StringUtils.safeFormat(
+                              "Server[%s] failed to complete request[%s] within timeout[%s millis].",
+                              serverId,
+                              config.getLoadTimeoutDelay().getMillis()
+                          )
+                      );
+                    }
+                  },
+                  config.getLoadTimeoutDelay().getMillis(),
+                  TimeUnit.MILLISECONDS
+              );
+            } else {
+              return curr;
+            }
+          }
+      );
+    }
+
+    public void requestCompleted(boolean failed, String failMsg)
+    {
+      synchronized (lock) {
+        Future cancellationFuture = cancellationFutureRef.get();
```

Comment:
I really don't understand this cancellation future stuff....

Reply:
this code is removed.